File: 240726/숫자 빠르게 찾기/find-number-fast.js
const main = () => {
    const input = require('fs').readFileSync(0).toString().split("\n");

    const arr = input[1].trim().split(" ").map(Number);
    const nums = input.slice(2).map(i => Number(i.trim()));

    nums.forEach(num => {
        console.log(binarySearch(num, arr));
    })
}

const binarySearch = (val, arr) => {
    let l = 0;
    let r = arr.length - 1;

    while (l <= r) {
        const mid  = Math.floor((l + r) / 2);

        if (arr[mid] === val) return mid + 1;
        
        if (val < arr[mid]) {
            r = mid - 1;
        } else {
            l = mid + 1;
        }
    }

    return -1;
}

main();

File: 240728/숫자 빠르게 찾기/find-number-fast.kt
import java.io.*

var arr: Array<Int> = emptyArray()

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val nums =
        br
            .readLine()
            .trim()
            .split(" ")
            .map { it.toInt() }
            .toTypedArray()
    arr =
        br
            .readLine()
            .trim()
            .split(" ")
            .map { it.toInt() }
            .toTypedArray()
    for (i in 0 until nums[1]) {
        search(br.readLine().toInt())
    }
}

fun search(target: Int) {
    var left = 0
    var right = arr.size - 1
    while (left <= right) {
        val mid = (left + right) / 2
        val num = arr[mid]
        if (num == target) {
            println(mid + 1)
            return
        } else if (num > target) {
            right = mid - 1
        } else {
            left = mid + 1
        }
    }
    println(-1)
}